### PR TITLE
[MM-53840] Add migration workflow

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
We are adding a new CI check for automatically validating the data migration from MySQL to PostgreSQL. This process 

The process looks like in a sequential order as following:
- Generate test data (We use `make test-data` target but we can also feed the MySQL DB with random data, see the [gist](https://gist.github.com/isacikgoz/19e28482e6676280e494cb87812d4da8) here.).
- Create a PostgreSQL schema and apply all of the migrations we already have via morph.
- Configure `pgLoader` and run.
- Use [dbcmp](https://github.com/mattermost/dbcmp) to compare database contents.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53840

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: .github/workflows/migration.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        run:
+          working-directory: server
+      steps:
+      - name: Checkout mattermost project
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - name: Setup Go
+        uses: actions/setup-go@4d34df0c2316fe8122ab82dc22947d607c0c91f9 # v4.0.0
+        with:
+          go-version: ${{ env.go-version }}
+          cache-dependency-path: server/go.sum
```
Comment (toninis) [inline .github/workflows/migration.yml:23]: Since you are running everything on docker you do not need to initialise go

File: .github/workflows/migration.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+           docker run --net ${COMPOSE_PROJECT_NAME}_mm-test \
+           --ulimit nofile=8096:8096 \
+           --env-file=build/dotenv/migration.env \
+           -v $(go env GOCACHE):/go/cache \
+           -e GOCACHE=/go/cache \
+           -v $PWD:/mattermost \
+           -w /mattermost \
+           $BUILD_IMAGE \
+           make test-migration
+      - name: Upload artifacts
```
Comment (toninis) [inline .github/workflows/migration.yml:53]: For some reason logs are not there after the run . There is a warning in the worfklow

File: .github/workflows/migration.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        with:
+            name: Migration logs
+            path: migration.log
+            retention-days: 7
+      - name: Stop docker compose
```
Comment (toninis) [inline .github/workflows/migration.yml:59]: this is not needed since the runner immediately dies after job execution is done. Unless you want to test the successful termination of the process this may result to false negative tests

File: .github/workflows/migration.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
+name: Database Migration Test
+on:
+  workflow_call:
+env:
+  go-version: "1.19.5"
```
Comment (toninis) [inline .github/workflows/migration.yml:5]: Also this is not needed probably

**Review by toninis — APPROVED** [review]
Worfklow approved ✅ 

Some minor comments 😄

File: .github/workflows/migration.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        run: |
+          docker run --net ${COMPOSE_PROJECT_NAME}_mm-test \
+            --ulimit nofile=8096:8096 \
+            --env-file=build/dotenv/migration.env \
+            -v $(go env GOCACHE):/go/cache \
+            -e GOCACHE=/go/cache \
+            -v $PWD:/mattermost \
+            -w /mattermost \
+            $BUILD_IMAGE \
+            make test-data
```
Comment (streamer45) [inline .github/workflows/migration.yml:34]: As you mentioned the key here will be to generate sufficiently diverse data to gain confidence that the migration is stressing most (if not all) tables/columns.

Do we have a ticket to plan improvements on this front? Regardless of its implementation (SQL functions vs server command or else) it would be good to prioritize this sort of work.

**Review by streamer45 — APPROVED** [review]
Very thorough, thanks!

File: .github/workflows/migration.yml (hunk 5 above)
Comment (isacikgoz) [reply .github/workflows/migration.yml:34]: Makes sense, I think it would be best to generate via the server app, to get a real life data. Created the ticket here: https://mattermost.atlassian.net/browse/MM-53979

File: server/Makefile [hunk 6]
```diff
@@ -828,3 +830,34 @@ migrations-extract:
 	@echo Listing migration files
 	@echo "# Autogenerated file to synchronize migrations sequence in the PR workflow, please do not edit.\n#" > channels/db/migrations/migrations.list
 	find channels/db/migrations -maxdepth 2 -mindepth 2 | sort >> channels/db/migrations/migrations.list
+
+test-migration:
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply server/Makefile:834]: I think we can increase the confidence by also generating the same data (through server, this time with a psql backend) and compare with the migrated data. I think that would be two way validation. The source of truth can be that (the data generated via server with psql backend) as well.

File: .github/workflows/migration.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        run:
+          working-directory: server
+      steps:
+      - name: Checkout mattermost project
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - name: Run docker compose
+        run: |
+          cd build
+          docker-compose --ansi never run --rm start_dependencies
+          cat ../tests/test-data.ldif | docker-compose --ansi never exec -T openldap bash -c 'ldapadd -x -D "cn=admin,dc=mm,dc=test,dc=com" -w mostest';
```
Comment (agnivade) [inline .github/workflows/migration.yml:21]: Do we really need ldap data?

File: server/Makefile [hunk 8]
```diff
@@ -828,3 +830,34 @@ migrations-extract:
...
+	@sed -i'' -e 's|{{ .pg_password }}|${PG_PASSWORD}|g' tests/temp.load
+	@sed -i'' -e 's|{{ .target_schema }}|${PG_SCHEMA}|g' tests/temp.load
+
+	# do for mysql credentials
+	@sed -i'' -e 's|{{ .mysql_user }}|${MYSQL_USERNAME}|g' tests/temp.load
+	@sed -i'' -e 's|{{ .mysql_password }}|${MYSQL_PASSWORD}|g' tests/temp.load
+	@sed -i'' -e 's|{{ .source_schema }}|${MYSQL_SCHEMA}|g' tests/temp.load
+
+	# install pgloader
+	apt install -y pgloader
```
Comment (agnivade) [inline server/Makefile:854]: Can we try to install `pgloader` by doing a `docker run` from the workflow file? We typically try to keep the Makefile commands to be platform insensitive.

File: server/Makefile [hunk 9]
```diff
@@ -828,3 +830,34 @@ migrations-extract:
...
 	@echo "# Autogenerated file to synchronize migrations sequence in the PR workflow, please do not edit.\n#" > channels/db/migrations/migrations.list
 	find channels/db/migrations -maxdepth 2 -mindepth 2 | sort >> channels/db/migrations/migrations.list
+
+test-migration:
+	$(GO) install github.com/mattermost/morph/cmd/morph@latest
+	# apply postgres migrations to the new db
+	bin/morph apply up --driver postgres --dsn "${POSTGRES_DSN}" --path ${POSTGRES_MIGRATIONS_PATH} --number -1
+
+	# create temporary file for the config
+	@cp tests/template.load tests/temp.load
```
Comment (agnivade) [inline server/Makefile:840]: Why is this needed? Can we not use the original template and modify that?

**Review by agnivade — COMMENTED** [review]
Some comments from my side. Really good to run this in CI to get validation.

File: .github/workflows/migration.yml (hunk 7 above)
Comment (isacikgoz) [reply .github/workflows/migration.yml:21]: Hmm, I think not for the test-data, it was copied form another workflow. I'll remove.

**Review by agnivade — APPROVED** [review]
